### ISSUE TITLE
Discussion on the general user interface in PyAbel

### ISSUE THREAD

**Issue by DanHickstein** [PyAbel/PyAbel#14]
In order to simplify and leave room for other transform methods, should we combine the basis set generation routines (basis.py) with the main basex code (core.py)? Some of the general image processing functions that are currently living in core.py (generate_speeds and all functions below that) should probably go into their own file since they will likely be common to many of the transform methods. 

Additionally, would it be better to allow the basex transform to just be one function call? Currently, there is one call to make/load the basis sets and another function to perform the transform. This might confuse most users, who just want to call a single function to perform a basex transform on their image. Why not just have one function that takes the image and does the transform?

**Comment by rth** [comment]
For the first part, I agree completely, well we can move all related part from `basis.py` and `core.py` to `basex.py` and more the rest under e.g. `utils.py`, `postprocessing.py` (for the `generate_speeds`), etc. 

As to the second point, I agree that we can merge loading of basis set and the BASEX calculation in one function call (although personally I find it more intuitive in 2), but there is also a number of other steps to consider that should be common to all inverse Abel transforms:

  1. Pre-processing (optional):
      - Possibly filtering: median, etc.
      - Centering the image
  2. Inverse Abel transform
     - in the case of BASEX in 1 step or 2 steps
  3. Post-filtering:
     - Cleaning up the reconstruction with Gaussian, median filters, etc.
  4. Post-processing depending on the application:
    - calculating speeds, etc.

So right now all of these steps are performed in 2 function calls, and I don't think it would help to merge them in 1 function call . I mean we can do a wrapper if somebody find that convenient, but in the work-flow it could be useful to separate these different steps as function calls.  I mean that's probably what would be done in Matlab as well. 

 No, what  do you think?

Say for basex what do you think about something roughly as follows,

     from abel.basex import BASEX

     # img is the input image
     b = BASEX(img, n=500, nbf=250, basis_dir='./', verbose=True)
     b.pre_process(center=(681,491), median_size=2, gaussian_blur=0)
     res = b.run()
     res2 = b.post_process(median_size=2) # optional
     IM = b.calculate_speeds() # or some other calculation depending on the application
     

where `calculate_speeds`, `pre_process`, and `post_process` would be common to all implementations. For a different implementation what would change is the first 2 lines:
   
     from abel.some_other_implementation import OnionPeeling
      
     b = OnionPeeling(img, some_other_parmaters)
     # the remaining steps are same as above

In addition, we could also create a wrapper independent of the implementation something like:
    
     from abel import InvAbel
     
     b = InvAbel(img, method='BASEX',
                                 options=dict(n=500, nbf=250, basis_dir='./'), verbose=True)
     # the remaining steps are same as above
     

 I think that would be more pythonic and modular than 

     res2, IM = BASEX(img, n=500, nbf=250, basis_dir='./', verbose=True, center=(681,491), 
                         median_size=2, gaussian_blur=0, post_median=0, use_speeds=True)
      
 as the latter that would require to think about lot of input parameters (and there is a risk of forgetting something), particularly as new features are added (i.e yet more input parameters), which could be  confusing for users.     

I mean that's just a very rough first proposal for the user interface in PyAbel, I don't know what interface you and @DhrubajyotiDas  think would be best? This may be a long term discussion, we don't have to change things right away...

**Comment by DanHickstein** [comment]
You make a lot of good points here, and, realistically, any of these options would probably be fine as long as we provide good examples of use.

But I think that most users are going to view the inverse Abel transform as just another image manipulation, and will want it to be accomplished in one line. They will want to specify the center and leave all the other values to default, so they can simply use: 

```python
recon, speeds = abel.BASEX(data,center=(250,250))
```

(The center function could even default to ``False``, in which case the center of the image would be used.) 

Sure, there would be lots of keywords in this single function, but most of them are fine left to their default values. The situation is similar to [scipy.ndimage.interpolation.rotate](http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.ndimage.interpolation.rotate.html#scipy.ndimage.interpolation.rotate) – there are numerous keywords, but 99% of users will just use ``rotate(data, 45)``.

The pre and post-processing median and gaussian filters are not really that important, and could almost be removed entirely, since they could be easily accomplished by the user (using the numpy functions) before or after calling the BASEX function. The argument for leaving them is that it's a nice suggestion to the users that they can use filters to improve their transforms in the case of noisy images. But, if we really wanted to minimize the keywords, we could get rid of them.

On another note, I wonder if we should include the functionality to just allow the user to specify ``n`` (letting ``nbf`` default to ``'automatic'`` or something), in which case ``nbf`` would be set to ``(n-1)/2``, which makes one basis function per pixel, which is the desired basis set in most cases. This allow the user to adjust the region of the image that is transformed without worrying about the details. 

Of course, if they want to tweak their basis set, they can still set ``nbf`` to whatever they want.

**Comment by DhrubajyotiDas** [comment]
While I agree that both scenarios have merit, I think we need to discuss whether we are designing a tool whose primary purpose is to do a fundamental mathematical procedure like the Abel inversion (using BASEX, POP, etc.) or something more *packaged* like the [scipy.ndimage.interpolation.rotate](http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.ndimage.interpolation.rotate.html#scipy.ndimage.interpolation.rotate) path where the end user has a most-likely usage-scenario that our design is geared towards.

I am biased towards the first, more fundamental, way of presenting these tools, such as the examples @rth provided, because my data processing pipeline was very different from what BASEX traditionally has been used for (no vertical symmetry, no speeds calculation, etc.).

Regarding specifying `n` with `nbf` defaulting to `automatic`  or `(n-1)/2`, seems like the only case where `nbf != (n-1)/2` is when the user is explicitly looking for additional smoothing from using fewer basis functions. I think setting `nbf` to a `default` value makes sense.

**Comment by DanHickstein** [comment]
Okay, maybe I do not clearly understand the motivation for using a syntax that requires numerous function calls. Is the idea to make things more modular? So that, for instance, users can generate basis sets without actually doing the transform, or complete the transform without the speed calculation? Or is the idea that we just want to make it very explicit to the user exactly what steps the algorithm is performing? 

I think that we are in a situation where 90% of the users are going to want the same thing: to center an image and perform the abel transform. Some will want the speed distribution and some will not, but we already toggle this with the ``calc_speeds`` keyword.  

I can't really imagine a situation where somebody would want to generate the basis sets without performing a transform. But, if they do, then that function is available to them in ``abel.basis.generate_basis_sets()``. Likewise, if all they want to do is to center the image, then they can use ``abel.core.center_image()``. 

Maybe a little restructuring of the code can allow both options to live together simultaneously. If we moved the loading of the basis sets into the main transform function (currently called ``basex_transform``), then we could move to a situation where most users could call:

```python
recon, speeds = abel.basex(data, center=(250,250), speeds=True)
```

which would be equivalent to the more modular option
```python
centered_data = abel.tools.center_image(data, center=(250,250))
recon = abel.basex.basex_transform(data, n=501, nbf=250)
speeds = abel.tools.generate_speeds(recon)
```
where this assumes that we moved all of the functions common to many transforms into a new file called ``tools``.

Does this satisfy both camps? Or is it the object-oriented behavior that's the most important in your view?

I'll make a separate issue for the ``nbf`` defaulting to ``(n-1)/2``

**Comment by rth** [comment]
@DanHickstein  I do get your  point from the first comment that user working with velocity-map-imaging (VMI) datasets in your field (no idea how it work ^^)  would want to simply call `basex(data, center=(250,250)) ` and get the result.  My concern is more to have access to elementary operators so this could be used more easily with a  different workflow, similarly to @DhrubajyotiDas . For instance, in my field, seeing `speeds=True` argument would be confusing for users because, when analyzing interferometry images of plasma flows, speed would refer to the flow velocity and they would not see what this had to do with an inverse Abel transform. So we would need to think how to make this sufficiently generic while satisfying expectations of different communities.  I don't have strong convictions for or against object-oriented behavior.   

I completely agree  with your second comment, that this would be probably the best way to handle it.  So for instance we could define 2 level of functions:

##### Low level functions (mostly for internal use, or advanced users)
that are completely self consistent and perform only one operation. As you mention, these can be:
  - centering function
  - generate speeds
  - BASEX: one internal function to generate basis sets, one internal function to perform the transform and one integrated function that does both in one call 
  - some preprocessing and post_processing: not critical, as these are performed by numpy but still nice to have as a shortcut

##### High level function (essentially integrated workflows)
mostly used in the examples, this may be domain specific or generic, might need to think about the specifics. 
   `basex(data, center=(250,250))`
I agree with that the only point I have an issue with is the integrated `speed`  argument because this is very specific to VMI datasets and would be confusing in a different context (and beside it is unclear what do to do then when this is applied to a generic inverse Abel implementation that does not have up/down symmetry). Or call this function  `basex_vmi` ?  And a `basex` function without the speed arguments?

But anyway, this is non critical, I'm sure we could discuss this at later phases and reach a compromise there. The situation would also be clearer when other Inv. Abel implementations are added, as to see how to make them work together.

**Comment by DanHickstein** [comment]
Okay, it sounds like we are converging. Low-level functions to satisfy the hackers, and a one-liner high level function for the chemists :smile:.

You are also right that ``speeds`` only the appropriate term when transforming momentum distributions (typically photoelectron/photoion momentum distributions), and the term is nonsensical for the general Abel transform. 

I am struggling for what the correct term is:
``angle integration in spherical coordinates``
``radial_intensity``
``radial_distribution``

Maybe just some simple keyword, like ``calc_radial`` and then we can explain in the docstring that this enables/disables the calculation of the one-dimensional radial distribution after transforming to spherical coordinates.

**Comment by DhrubajyotiDas** [comment]
While we are at it, would there be interest in discussing the up/down symmetry aspect of the transforms? As it stands, those are a special case of Abel forward-transformed distributions. The only requisite symmetry for the BASEX technique to work is the lateral `f(-r) = f(r)` symmetry. 

The code in [PyAbel](https://github.com/PyAbel/PyAbel) currently presupposes `f(r,-z) = f(r,z)` symmetry, which seems to be a feature of the photoelectron/photoion measurements (?) and is leveraged as a computational time-saver during basis-set generation, but at the expense of generality. 

Most of my work in [pyroBASEX](https://github.com/DhrubajyotiDas/pyroBASEX) has been to generalize the [PyAbel](https://github.com/PyAbel/PyAbel) code to enable handling situations where the raw data just has `f(-r) = f(r)` symmetry. This is done without significant increases in basis-set generation computational time because the additional computations necessary to generate `vertical` basis functions take much less time compared to the `horizontal` basis functions, as the `vertical` Gaussian basis functions don't need to be forward Abel transformed.

If such a feature were to be incorporated into PyAbel, it would have minimal impact on the end-user experience when compared to the situation now. For example, if the user does not specify `n_vert` or `nbf_vert` explicitly, or if `n_vert == n_horz` etc., it would work as usual. If the provided `raw_data` is not square, or if `n_vert != n_horz`, then two sets of basis functions in either direction could be calculated, and when the user invokes the actual reconstruction, the specified center would be a full column instead of a single point.

Thoughts?

**Comment by rth** [comment]
Regarding  merging pyroBASEX into PyAbel, I don't know what would be the best approach. A proposition could be for instance to make 2 independent basis set generation routines one for each (as it's already hard enough to understand what it is doing without adding more conditional tests). Then from a user perspective, in some higher order function say `basex_get_basis_sets_cached` (in PR above) and `abel.basex.BASEX`,  have a single `n` argument that will be integer for the regular BASEX  transform, and tuple `(n_vert, n_horz)` for pyroBASEX type of transform. This way you can do the same thing without introducing more input argument, and it will probably be clearer from the user perspective. Most of the other code can be reused no? or maybe adding just a few tests to differentiate between the 2 use cases (e.g. in the scaling factor that should be put in a different location). What do you think?

**Comment by DhrubajyotiDas** [comment]
>a single `n` argument that will be integer for the regular BASEX transform, and tuple `(n_vert, n_horz)` for pyroBASEX type of transform

Yes, that's basically it. 

I also agree that it makes sense to do two independent basis set generation routines. It's slightly redundant, but in my opinion the gain in code legibility makes it worth it. Most of the pyroBASEX code can be reused, although some utility functions like the one that centers the input `raw_data` might need to be treated with some care to fit with the larger structural changes to the codebase. 

Regarding the scaling factor, we could either multiply it to the `Ci` matrix (for now), or a more robust implementation where the scaling factor is determined from the unit tests each time (since doesn't take that long to run the unit tests each time a new basis set is generated, I think), in which case it can be placed wherever it currently is in PyAbel.

**Comment by DanHickstein** [comment]
I support the idea of enabling the use of up/down-asymmetric images. In photoelectron spectroscopy, our images are theoretically up/down symmetric, and so a lot of reconstruction algorithms make this approximation in order to gain better signal-to-noise. However, there are many scenarios where the up/down asymmetry is broken – for instance, in the Xenon example that I provided, there is a hot spot on the bottom left side of the detector and it would be nice to complete the transform only on the upper half on the image.

Let me see if I understand correctly: it's just the basis set generation and the scaling factor that are different? And ``n`` would need to be ``(n_vert, n_horiz)``. 

And @rth is proposing that the user differentiate between the two transforms simply by using an integer or a tuple for n? I suppose this is okay as long as it's documented. Thought it's a little weird that ``n=500`` and ``n=(500,500)`` produce different behavior. Adding a ``vertical_symmetry=True`` keyword would be the alternative...

**Comment by rth** [comment]
You are right, adding an explicit key word `vertical_symmetry=True`, in addition to what was mentioned,  would probably be more intuitive. Then say  `vertical_symmetry=False` and `n=500` can be translated as `n=(500, 500)`, while `vertical_symmetry=True` and `n=(500, 500)` should raise an error.

**Comment by DhrubajyotiDas** [comment]
>Thought it's a little weird that `n=500` and `n=(500,500)` produce different behavior.

Assuming we always invoke the pyroBASEX approach whenever `n = <tuple>`, from the user's perspective, `n=500` and `n=(500,500)` would provide the exact same reconstruction. We could set up a check for that where if the two values in the tuple are the same, treat it as an integer and proceed with PyAbel.

`vertical_symmetry=True` would definitely work, but at the cost of introducing yet another parameter.

**Comment by rth** [comment]
@DhrubajyotiDas Yes, so just to make I understood everything. At present pyroBASEX with `n_vert, n_horiz = 500, 500` would produce different results than BASEX with `n=500`, right?  Then if this is true, it possible to do make `n=500` and `n=(500,500)` provide the exact same reconstruction in PyAbel. However this would mean that `n=500, 500` and `n=500, 498` would be significantly different as one has the up down symmetry constrain and not the other. In which case an additional argument such as `vertical_symmetry=True` would make things more clear.  Actually, I don't understand: is there a link between the fact that the image is square and the up/down symmetry in BASEX/pyroBASEX? Can you not have an image `n=(500,500)` without the up/down symmetry?

**Comment by DanHickstein** [comment]
Was about to ask this same question. I would assume up/down symmetry is irrelevant of dimensions...

**Comment by DhrubajyotiDas** [comment]
>At present pyroBASEX with `n_vert, n_horiz = 500, 500` would produce different results than BASEX with `n=500`, right?

Yes, they should produce different results. Now that I think about it, the only time pyroBASEX and PyAbel will have the same output in this situation is if the `raw_data` being transformed is exactly symmetric (no hotspots, etc. like @DanHickstein mentioned). For real world data, that's never going to be true, and PyAbel will "average" the two halves to provide better signal/noise while pyroBASEX will leave them untouched. In that case, specifying `vertical_symmetry=True` makes a material difference.

>I would assume up/down symmetry is irrelevant of dimensions...

I guess I was confused by the case where `raw_data` is rectangular and `vertical_symmetry=True`. It didn't occur to me that in this case, with PyAbel, `raw_data` can be enlarged to a square with side = the larger of the two dimensions of `raw_data` by padding zeroes, and then inverted. 

So to recap

* `vertical_symmetry=True` -> Use PyAbel
* `vertical_symmetry=False` or `unspecified` -> Use pyroBASEX

Examples

1)  `n=500` -> Use pyroBASEX, `n_horz, n_vert = 500,500`
2)  `n=500, vertical_symmetry=True` -> Use PyAbel as usual
3)  `n=(500,500), vertical_symmetry=True` -> Same as above, convert `n` from tuple to integer `n=500`
4)  `n=(500,500)` -> Same as 1)
5) `n=(500,498)` -> Use pyroBASEX, `n_horz, n_vert = 500,498`
6) `n=(500,498), vertical_symmetry=True` -> Enlarge `raw_data` to `(500,500)`, then proceed as 2)

**Comment by DanHickstein** [comment]
This seems like a good user-interface concept. Can you make a separate issue detailing the implementation of the asymmetric case (basically just summarize what we said here)? 

I think that it would be a good idea to implement the other user-interface changes first (then we can close this issue), and then we can go ahead and work on implementing the asymmetric basis sets.

**Comment by DanHickstein** [comment]
So, I think that the UI now operates as we discussed. Should we close this?
